### PR TITLE
Modal with options

### PR DESCRIPTION
**Pull request by enahum**
#### Summary
This adds the ability to show a modal with options and in this case adds the modal for a long press on the channel drawer items to show different options for each channel

#### Ticket Link
No ticket assign

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [ ] Added or updated unit tests (required for all new features)
- [ ] All new/modified APIs include changes to the drivers
- [x] Has UI changes
- [x] Includes text changes and localization file updates

### CODE REVIEW

File: app/components/channel_drawer/channel_list.js [hunk 1]
```diff
@@ -147,22 +150,146 @@ class ChannelList extends React.Component {
...
+                defaultMessage: 'Are you sure you want to leave the {term} with {name}?'
+            }, {
+                term: term.toLowerCase(),
+                name: channel.display_name
+            }),
+            [{
+                text: formatMessage({id: 'mobile.channel_list.alertNo', defaultMessage: 'No'})
+            }, {
+                text: formatMessage({id: 'mobile.channel_list.alertYes', defaultMessage: 'Yes'}),
+                onPress: () => {
```
Comment (jarredwitt) [inline app/components/channel_drawer/channel_list.js]: If a user chooses to leave a channel that they currently have open, the channel stays open in the channel view after they leave. We should move the user to one of their other channels right?

File: app/components/channel_drawer/channel_list.js [hunk 2]
```diff
@@ -336,8 +463,27 @@ class ChannelList extends React.Component {
             );
         }
 
+        const modalOptions = (
+            <ModalOptions
+                title={this.state.modalTitle}
+                options={this.state.options}
+                visible={this.state.showOptions}
+                onChange={(option) => {
```
Comment (jarredwitt) [inline app/components/channel_drawer/channel_list.js]: The ModalOptions component re-renders once when the drawer is opened and then twice when a new channel is selected. Extending PureComponent did not change this behavior until this function was moved to a class method. Pretty small component to re-render, but just wanted to point it out if we're looking to make performance opts down the road.

File: app/components/channel_drawer/channel_list.js (hunk 1 above)
Comment (enahum) [inline app/components/channel_drawer/channel_list.js]: weird it suppose to go the the `DEFAULT_CHANNEL` I'll take a look

File: app/components/channel_drawer/channel_list.js (hunk 2 above)
Comment (enahum) [inline app/components/channel_drawer/channel_list.js]: you're saying moving this from a Component to a function?

Comment (jarredwitt) [inline app/components/channel_drawer/channel_list.js]: Oh wow, bad comment now that I look at it. I meant to add that it didn't stop re-rendering until I extended PureComponent and moved the function that is in the onChange property to a method on channel_list.

**Comment by enahum** [comment]
@jarredwitt should be fixed now, and thanks for submitting that PR against this one

**Comment by enahum** [comment]
@hmhealey PR rebased

File: app/components/channel_drawer/channel_list.js [hunk 3]
```diff
@@ -287,6 +399,17 @@ class ChannelList extends React.Component {
         return data;
     };
 
+    modalOnChange(option) {
+        this.props.handleDisableDrawer(false);
+        if (option) {
+            option.action();
+        } else {
+            this.setState({
```
Comment (hmhealey) [inline app/components/channel_drawer/channel_list.js]: I think you have every option explicitly closing the modal, so you could change this to
```
this.setState({showOptions: false});

if (option) {
    option.action();
}
```
and get rid of that duplicated code above.

File: app/components/modal_options.js [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+        color: '#7f8180'
+    }
+});
+
+export default class ModalOptions extends PureComponent {
+    static propTypes = {
+        title: PropTypes.string,
+        options: PropTypes.array,
+        visible: PropTypes.bool,
+        onChange: PropTypes.func,
```
Comment (hmhealey) [inline app/components/modal_options.js]: Instead of `onChange` could you call this something like `onOptionSelected`? It's currently not clear what it's supposed to be used for

File: app/scenes/channel_drawer/channel_drawer.js [hunk 5]
```diff
@@ -42,8 +55,12 @@ export default class ChannelDrawer extends React.PureComponent {
                 channelMembers={channelMembers}
                 theme={theme}
                 onSelectChannel={this.selectChannel}
-                onViewChannel={this.props.actions.viewChannel}
-                handleCloseDM={this.props.actions.closeDMChannel}
+                onViewChannel={viewChannel}
```
Comment (hmhealey) [inline app/scenes/channel_drawer/channel_drawer.js]: This would be a separate PR, but I think we need to consider making the `ChannelList` or its children bind these actions themselves. `ChannelDrawer` is ending up very tightly coupled to `ChannelList` which is something we were hoping redux would remove the need for

File: app/components/channel_drawer/channel_list.js [hunk 6]
```diff
@@ -143,24 +148,131 @@ class ChannelList extends React.Component {
...
     };
 
+    onShowModal = (channel) => {
+        const {formatMessage} = this.props.intl;
+        let open;
+        let close;
+        let favorite;
+        let modalTitle;
+
+        this.props.handleDisableDrawer(true);
```
Comment (hmhealey) [inline app/components/channel_drawer/channel_list.js]: Maybe make this automatically happen when a `ModalOptions` is displayed? I can't think of a case where we'd want the drawers available in the modals

File: app/reducers/navigation/index.js [hunk 7]
```diff
@@ -16,7 +16,8 @@ const initialState = {
     leftDrawerOpen: false,
     leftDrawerRoute: null,
     rightDrawerOpen: false,
-    rightDrawerRoute: null
+    rightDrawerRoute: null,
+    leftDrawerDisabled: false
```
Comment (hmhealey) [inline app/reducers/navigation/index.js]: Can you make this into `drawersDisabled`? I can't think of any circumstances when you'd want one disabled and not the other

File: app/components/channel_drawer/channel_item.js [hunk 8]
```diff
@@ -152,37 +151,14 @@ export default class ChannelItem extends React.Component {
             }
         }
 
-        let closeButton = null;
```
Comment (enahum) [inline app/components/channel_drawer/channel_item.js]: Yeah I forgot to add it back, will do

File: app/components/channel_drawer/channel_list.js (hunk 6 above)
Comment (enahum) [inline app/components/channel_drawer/channel_list.js]: Yeap will do

File: app/components/channel_drawer/channel_list.js [hunk 9]
```diff
@@ -143,24 +148,131 @@ class ChannelList extends React.Component {
...
+
+        if (channel.type === Constants.DM_CHANNEL) {
```
Comment (enahum) [inline app/components/channel_drawer/channel_list.js]: Actually I want to leave `<ModalOptions>` to receive the options as props as that component shouldn't be related only to channels, or I'm not getting you here

File: app/components/channel_drawer/channel_list.js [hunk 10]
```diff
@@ -143,24 +148,131 @@ class ChannelList extends React.Component {
...
+                },
+                text: formatMessage({id: 'channelHeader.addToFavorites', defaultMessage: 'Add to Favorites'})
+            };
+        }
+
+        const options = [open, favorite];
+        if (close) {
+            options.push(close);
+        }
+        this.setState({options, modalTitle, showOptions: true});
```
Comment (enahum) [inline app/components/channel_drawer/channel_list.js]: Same here

File: app/components/channel_drawer/channel_list.js (hunk 3 above)
Comment (enahum) [inline app/components/channel_drawer/channel_list.js]: I'll review it

File: app/components/modal_options.js (hunk 4 above)
Comment (enahum) [inline app/components/modal_options.js]: Yeap will do

File: app/scenes/channel_drawer/channel_drawer.js (hunk 5 above)
Comment (enahum) [inline app/scenes/channel_drawer/channel_drawer.js]: I can do it in this PR, not a problem

File: app/reducers/navigation/index.js (hunk 7 above)
Comment (enahum) [inline app/reducers/navigation/index.js]: sure thing, makes total sense. Will do

File: app/components/channel_drawer/channel_list.js (hunk 3 above)
Comment (enahum) [inline app/components/channel_drawer/channel_list.js]: Ok I'm going to leave this one as it is cause in the event that the `leave channel` is selected which shows later on an alert, there is a weird behavior that makes the alert shown in the background but blocking the screen which makes the app unusable, this way its handled correctly, I'll add a comment in the code

**Comment by enahum** [comment]
@hmhealey take another look

File: app/components/modal_options/modal_options.js [hunk 11]
```diff
@@ -0,0 +1,223 @@
...
+        borderBottomColor: '#999999'
+    },
+
+    titleTextStyle: {
+        fontSize: 15,
+        color: '#7f8180'
+    }
+});
+
+export default class ModalOptions extends PureComponent {
```
Comment (hmhealey) [inline app/components/modal_options/modal_options.js]: Sorry, I know there's a bunch of renaming that you'll have to do for this, but can you rename this to `OptionsModal`? Every time I see this, I keep thinking that it's something you pass into a modal to tell it how to look and behave. I think that's also consistent with modal names in the webapp and with things like containers that are `PostListContainer` instead of `ContainerPostList`

Comment (hmhealey) [inline app/components/modal_options/modal_options.js]: Sorry, I know there's a bunch of renaming that you'll have to do for this, but can you rename this to `OptionsModal`? Every time I see this, I keep thinking that it's something you pass into a modal to tell it how to look and behave. I think that's also consistent with modal names in the webapp and with things like containers that are `PostListContainer` instead of `ContainerPostList`

File: app/navigation/router.js [hunk 12]
```diff
@@ -148,15 +161,18 @@ class Router extends React.Component {
                         render={this.renderTransition}
                         configureTransition={this.configureTransition}
                     />
+                    <ModalOptions/>
```
Comment (hmhealey) [inline app/navigation/router.js]: I assume this works correctly, but this actually renders the modal in the correct location even when drawers are open? I would've figured it would move the modal with the rest of the scene's contents

File: assets/base/i18n/en.json [hunk 13]
```diff
@@ -959,6 +959,8 @@
   "calling_screen": "Calling",
   "center_panel.recent": "Click here to jump to recent messages. ",
   "channel_header.addMembers": "Add Members",
+  "channelHeader.addToFavorites": "Add to Favorites",
+  "channelHeader.removeFromFavorites": "Remove from Favorites",
```
Comment (hmhealey) [inline assets/base/i18n/en.json]: Something else we should figure out soon is how the localization strings will get synced between the repos

File: app/navigation/router.js (hunk 12 above)
Comment (enahum) [inline app/navigation/router.js]: this is working exactly as it should

File: assets/base/i18n/en.json (hunk 13 above)
Comment (enahum) [inline assets/base/i18n/en.json]: This is really easy I already have it figured out, we can talk about this on one of the meetings
